[나무] TodoList with Ajax

wooooooood opened the pull request:
# 2주차. TodoList with Ajax
오늘 세션이 있는 날이라 구현한데까지 PR합니다!
어제 저녁에 페어프로그래밍을 통해 가이드를 많이 받아서 배운 내용 반영해서 추가커밋 예정입니다 😄

## 🎯 요구사항
- [x] todoItem 불러오기
- [x] todoItem 추가하기
- [x] todoItem 삭제하기
- [x] todoItem complete하기
- [x] todoItem contents 내용 수정하기
- [x] user별 투두리스트 불러오기

## 🎯🎯 심화 요구사항
- [ ] 데이터를 불러오기전 로딩바를 이용해, 사용자가 데이터가 불러와지고 있다는 것을 보여줍니다.
- [ ] fetch api 사용하는 부분을 async await을 사용하여 리팩토링합니다.
- [ ] github issue에서 라벨을 붙이는 것처럼, 우선순위에 따라서 badge를 추가합니다.
--> 렌더링하지 않았던 부분 추가
- [x] ES6 impot & export를 이용해 자바스크립트 파일을 리팩토링한다.

wmakerjun committed:
refactor(typo): fix typo

- conflict 처리

wmakerjun committed:
feat(priority): add chip selection to show priority

wooooooood committed:
todoItem 추가하기, 일부 module 적용

wooooooood committed:
todoItem 불러오기

wooooooood committed:
migrate. mission1

wooooooood committed:
Render chip selection

wooooooood committed:
Get user list

- merge upstream

wooooooood committed:
Toggle user selection, doesn't show user's list

wooooooood committed:
Delete todo item

ganeodolu commented on js/TodoInput.js:
```diff
@@ -1,15 +1,17 @@
+export default function TodoInput ($todoInput, onAdd) {
+  this.$todoInput = $todoInput
+  this.$todoInput.addEventListener('keyup', (e) => {
+    if (e.keyCode === KEY.ENTER_KEY) {
```
MDN에서 keyCode를 더이상 추천하지 않는다고 하여 앞으로 key를 사용하시는 것을 추천드려요 :frog:
https://developer.mozilla.org/ko/docs/Web/API/KeyboardEvent/keyCode
https://developer.mozilla.org/ko/docs/Web/API/KeyboardEvent/key

wooooooood replied:
아하 리뷰 감사합니다!! 왜 다른분들 코드 보면서도 의문을 갖지 못했던 부분일까요 ^_ㅠ 반영해서 수정하겠습니당~

2SOOY commented on js/TodoInput.js:
```diff
@@ -1,15 +1,17 @@
+  this.$todoInput = $todoInput
+  this.$todoInput.addEventListener('keyup', (e) => {
+    if (e.keyCode === KEY.ENTER_KEY) {
+      if (!/\S/.test(this.$todoInput.value) || this.$todoInput === null) {
```
안녕하세요! @wooooooood 님 :)

`(!/\S/.text(this.$todoInput.value))` 이 부분에 대해서 설명해주실 수 있나요 ?

wooooooood replied:
입력받은 문자열이 비었는지 검증해주는 정규식을 복붙..해온 부분이었는데요 ㅎㅎ; 말씀 듣고보니 그냥 봐서는 이해할수 없는 코드네욤.. 검증방식을 변경하도록 하겠습니당

ganeodolu replied:
`/ /` 정규식을 사용하여 
`(this.$todoInput.value)`의 값에 
`\S` 공백이 아닌 문자를
`!` 포함하지 않는지 
`.test` true 또는 false로 알려준다는 의미로 보시면 될 것 같아요. 🐸

wooooooood committed:
Done in Pair programming

- Implement Toggle
- Implement Edit
- UserTitle will change depending on active user
- Match code convention
- Getting Active user's todo list ..ing

2SOOY commented on js/TodoList.js:
```diff
@@ -15,32 +12,33 @@ export default function TodoList($todoList, data, removeItem) {
   this.bindEvents = () => {
-    document.querySelectorAll('.todo-item').forEach($item => {
+    document.querySelectorAll('.todo-item').forEach(($item) => {
       $item.querySelector('input.toggle').addEventListener('click', (e) => {
-        e.stopPropagation() //이벤트 버블링 막기
+        e.stopPropagation()
         const $todoItem = e.target.closest('.todo-item')
+        const { index } = $todoItem.dataset
 
         if ($todoItem.classList.contains('completed')) {
           $todoItem.classList.remove('completed')
+          this.data[index].isCompleted = false
+          
         } else {
           $todoItem.classList.add('completed')
+          this.data[index].isCompleted = true
         }
+        this.toggle($todoItem.id)
```
todo Item을 완료여부를 `toggle`하는 부분인데요 !

**1.** `Event delegation`을 적용하면 좋을 것 같아요 !
 현재는 TodoList 내부의 `모든 todo item`에 대해서 `click이벤트를 등록`하고 있습니다.
 이보다 todo item의 상위 개념인 TodoList에게만 click이벤트를 등록하고 `e.target.closest('li')`을 통해 todo item을 가져올 수 있을 것 같아요 :) 
 그러면 1번의 이벤트 등록만으로 모든 todo item의 이벤트 등록이 필요없어집니다 😄 
```js
this.$todoList.addEventListener('click', (e) => { // todoList에 click 등록
  if (e.target.className === 'toggle') {
   const todoItem = e.target.closest('li'); // toggle이 클릭된 todo Item 
   // toggle 로직 ...
 }
})
```
 

**2.** `TodoList`컴포넌트가 가진 this.data라는 `state를 변경하는 시점`
나무님의 state 변경 로직을 살펴보면,

`브라우저` 토글 버튼 클릭 
-> `client` todo list의 this.data의 해당 item의 isCompleted 변경 
-> toggle API 요청 -> `server` 저장된 todo list의 해당 todo Item의 isCompleted 변경
순으로 동작합니다.

 이는 client의 state를 미리 변경시키고, `서버로의 api요청이 정상적으로 이뤄지지 않았을 때`,
 `브라우저상에는 정상`적으로 토글이 동작한 것 같지만
 `실제 서버에서는 이를 반영하지 못하는 경우`가 발생할 수 있을 것 같아요 :)

 따라서 client의 state를 먼저 변경시키지 않고, 
서버에 토글 api이벤트를 요청한 후 변경된 서버의 state를 가져오는 식의 흐름으로 구성하면 좋을 것 같아요 😄
 
 요약하자면, 다음과 같은 로직으로 구성됩니다 :)
```
(브라우저) 토글 동작 
=> (클라이언트) toggle api요청 => (서버) state 변경 
=> (클라이언트) GET todoList api 요청 
=> (클라이언트) 받아온 todoList의 state에 반영
=> (브라우저) 해당 state에 맞게 렌더링
```

wooooooood replied:
핫 너무 상세한 리뷰 감사드립니다 ㅠㅠ
EventDelegation은 아직 직접 구현해본적이 없었지만 수연님이 샘플로 작성해주신것 감사히 참고하겠습니다!
그리고 api관련 실행 순서에 대해서도 꼼꼼히 짚어주셔서 감사드려요👍 정말 정상동작에만 집중하면 보이지 않는 중요한 디테일인것 같습니다! 앞으로의 api 작성이 더 강화될것 같습니다 ㅎㅎㅎ~~🙌감사합니당

ganeodolu commented on js/TodoList.js:
```diff
@@ -1,12 +1,9 @@
-export default function TodoList($todoList, data, removeItem) {
-  this.$todoList = $todoList
-  this.data = data
+import { KEY, ADDRESS } from './constants.js'
```
미션에는 없지만 짝프로그래밍할때 editing 모드가 동시에 여러 곳에 적용되는 것을 보았는데요. 
focusout 이벤트를 사용하여 editing 모드에서 focus를 잃어버렸을 때 다시 돌아가는 기능을 구현해주신다면 더욱 좋을 것 같습니다. :frog:

wooooooood replied:
아하 넵! 해당 이벤트를 이런 경우에 사용하는 것이었군요 ㅎㅎㅎㅎ 자세히 리뷰해주셔서 감사합니다~~

wooooooood committed:
Show active user's todo list

ganeodolu commented on js/TodoUser.js:
```diff
@@ -1,12 +1,52 @@
+    document.querySelectorAll('.ripple').forEach(($item) => {
+      $item.addEventListener('click', (e) => {
+        e.stopPropagation()
+        const $oldUser = e.currentTarget.parentNode.querySelector('.active')
+        $oldUser.classList.remove('active')
```
render할때 this.userName과 name이 같을 때에만 class에 active가 추가하되기 때문에 굳이 remove를 하지 않아도 될 것 같습니다. :frog:

ganeodolu commented on js/TodoUser.js:
```diff
@@ -1,12 +1,52 @@
+  this.bindEvents = () => {
+    document.querySelectorAll('.ripple').forEach(($item) => {
+      $item.addEventListener('click', (e) => {
+        e.stopPropagation()
```
혹시 여기서 이벤트 전파를 막는 이유는 무엇인지 궁금합니다 :frog:

wooooooood replied:
사실 저도 모르게 습관적으로? 어떤 전파가 있을지 몰라서 사용한 부분인 것 같네요..😭 다시 확인해보고 필요없는 코드라면 빼도록 하겠습니다!!

ganeodolu commented on js/TodoUser.js:
```diff
@@ -1,12 +1,52 @@
+
+    let result = ''
+    this.userList.map(({ _id, name }) => {
+      result += `<button class="ripple ${this.userName == name ? 'active' : ''}">${name}</button>`
```
비교시에는 `===` 를 사용하시는 것을 추천드려요

wooooooood replied:
핫 넵!! 당연한 부분을 짚어주셨네요 ㅠㅠ 감사합니다✨

ganeodolu commented on js/TodoUser.js:
```diff
@@ -1,12 +1,52 @@
+    document.querySelectorAll('.ripple').forEach(($item) => {
+      $item.addEventListener('click', (e) => {
+        e.stopPropagation()
+        const $oldUser = e.currentTarget.parentNode.querySelector('.active')
+        $oldUser.classList.remove('active')
+
+        const $selectedUser = e.target
+        this.userName = $selectedUser.innerText
+        $selectedUser.classList.add('active')
```
this.userName과 name을 비교하여 render하기 때문에 remove와 add를 사용하지 않아도 될 것 같습니다. :frog:

wooooooood replied:
헉.. 여기서는 active설정을 안해줘도 되는거였네요!! 햐 감사합ㄴㅣ다

ganeodolu commented on js/TodoUser.js:
```diff
@@ -1,12 +1,52 @@
+  }
+
+  this.bindEvents = () => {
+    document.querySelectorAll('.ripple').forEach(($item) => {
```
@2SOOY 님이 남겨주신 리뷰와 같이 여기서도 이벤트 위임을 사용해보는 것은 어떨까요?? 
```javascript
    this.$userList.addEventListener('click', (e) => {
        const $selectedUser = e.target
        this.userName = $selectedUser.textContent
        if(!$selectedUser.classList.contains('ripple')) return;
        this.setState(this.userName)
    })
```
그리고 `innerText` 보다 `textContent`를 사용하시는 것을 추천드려요
https://developer.mozilla.org/ko/docs/Web/API/Node/textContent#innerText%EC%99%80%EC%9D%98_%EC%B0%A8%EC%9D%B4%EC%A0%90

wooooooood replied:
아하! 링크 첨부까지 감사드립니다! 겸사겸사 헷갈렸던 부분도 정리되었습니다 ㅎ_ㅎ

ganeodolu commented on js/TodoUser.js:
```diff
@@ -1,12 +1,53 @@
 
-    this.render()
+  this.getUsers = () => {
+    fetch(`${ADDRESS.BASE_URL}/api/u`, {
```
fetch는 서버에서 가져오면서 실패할 수 있으므로 try ~ catch 등을 이용하여 예외처리를 해주면 더 좋을 것 같습니다.
https://helloworldjavascript.net/pages/290-exception.html

wooooooood replied:
넵!! 그러면 예외처리하는 연습을 먼저 하고 async await리팩토링을 해보겠습니다🔥

eastjun-dev commented on js/TodoUser.js:
```diff
@@ -1,12 +1,53 @@
+
+    let result = ''
+    this.userList.map(({ _id, name }) => {
+      result += `<button class="ripple ${this.userName == name ? 'active' : ''}">${name}</button>`
```
삼항연산자를 쓰실 때 false한 경우를 ''로 하는 경우에는
`${this.userName == name && 'active'}`
위와 같이 줄여서 표현하는게 가능합니다 😄

wooooooood replied:
오 뭔가 재미있는 표현같아요!! 리뷰감사드립니닷!!👍

eastjun-dev commented on js/TodoList.js:
```diff
@@ -1,46 +1,49 @@
+        this.toggle($todoItem.id)
       })
 
       $item.querySelector('button.destroy').addEventListener('click', (e) => {
```
가능하면, 콜백함수로 넘겨주는 부분을 익명함수로 작성하는게 아니라, 따로 분리하면 어떨까요? 익명함수로 구현해도 동작하는데는 무리가 없지만, 이벤트가 발생했을 때 무슨 동작을 해야하는지는 코드를 전부 봐야 알 수 있을 것 같아요. 협업 측면에서 봤을 때
`onDestroyHanlder`와같은 함수로 분리해서 구현한다면, 내부 동작까지 읽지 않아도, 이 이벤트 리스너가 어떤 콜백함수와 연결되어있는지 한눈에 파악하기 좋을 것 같아요 😄

wooooooood replied:
아하.. 더 협업에 가까운 구조네요! 분리해서 의미를 명확히 하는 연습! 하겠습니다 감사합니다~

eastjun-dev commented on js/TodoUser.js:
```diff
@@ -1,12 +1,53 @@
+    document.querySelectorAll('.ripple').forEach(($item) => {
+      $item.addEventListener('click', (e) => {
+        e.stopPropagation()
+        const $oldUser = e.currentTarget.parentNode.querySelector('.active')
```
`parentNode`로 찾으면, 다른 마크업이 추가되는 경우 에러가 발생할 수 있을 것 같아요. `closest`로 명확하게 element를 찾는게 더 좋을 것 같아요 😄

wooooooood replied:
핫 그런 경우에 문제가 되겠군요 짚어주셔서 감사합니당😃

eastjun-dev reviewed (changes requested):
나무님~! 늘 꾸준히 열심히 하시는 모습 덕분에 저도 늘 동기부여 얻고, 본받는것 같습니다 ㅎㅎ
이미 다른 분들이 자세한 피드백을 많이 남겨주셔서 제가 남길 부분이 별로 없었네요 ㅎㅎ
3번째 미션에서 더 꼼꼼한 피드백으로 찾아뵐게요 😄

CianJS commented on js/TodoList.js:
```diff
@@ -1,46 +1,49 @@
-        const {
-          index
-        } = e.target.closest('.todo-item').dataset
+        const { index } = e.target.closest('.todo-item').dataset
```
이렇게도 index를 얻어올 수 있었군요!
참고토록 하겠습니다. 감사합니다 ㅎㅎ

CianJS commented on js/TodoList.js:
```diff
@@ -50,49 +53,89 @@ export default function TodoList($todoList, data, removeItem) {
+  this.toggle = (_id) => {
+    fetch(`${ADDRESS.BASE_URL}/api/u/${this.userName}/item/${_id}/toggle`, {
+      method: 'PUT'
+    }).then(() => {
```
then을 통해 성공 케이스에 대한 따른 특별한 차이가 없다면 사용하지 않아도 괜찮을 것 같아요! 😄

wooooooood replied:
넵 불필요한 부분이 맞는것같습니당! 감사드려욧🙏

CianJS commented on js/TodoList.js:
```diff
@@ -50,49 +53,89 @@ export default function TodoList($todoList, data, removeItem) {
     })
   }
 
   this.get = () => {
```
저만 그런 것일 수 있는데 그저 `get`이라고 하면 코드의 내용을 읽어봐야만 역할을 파악할 수 있을 것 같아요!
`getTodoItems`같은 이름은 어떠신가요? 😄

wooooooood replied:
제가 api요청 방법을 그대로 따라간 네이밍이네요..^^ 리뷰감사드려욧!

wooooooood committed:
Do not use deprecated keyCode

wooooooood committed:
feedback (1)

lsw1164 commented on js/TodoList.js:
```diff
@@ -1,98 +1,141 @@
+      })
+    }).then(() => {
+      this.get()
     })
```
get, post, edit, delete, toggle 등
api 관련 코드들을 묶어놓은 스크립트를 따로 작성하시면 더 가독성에 도움이 될거같아요!

wooooooood replied:
아하 넵! 리뷰 감사합니다~ 분리하는 연습.. 더 열심히 하겠습니다! 👍

lsw1164 reviewed (commented):
코드 잘봤습니다!  짧은 피드백 남겼어요 :)
